src/tasks/LibraryBuilder/Templates/autoinit.c

```diff
@@ -0,0 +1,162 @@
+
+    // add extensions if required.
+    strlcpy (filename, name, sizeof (filename));
+    if (!has_extension) {
+        strlcat (filename, ".dll", sizeof (filename));
+    }
+
+    if (culture && strcmp (culture, ""))
+        res = snprintf (path, sizeof (path) - 1, "%s/%s/%s", bundle_path, culture, filename);
+    else
+        res = snprintf (path, sizeof (path) - 1, "%s/%s", bundle_path, filename);
+    assert (res > 0);
+
+    struct stat buffer;
+    if (stat (path, &buffer) == 0) {
+        MonoAssembly *assembly = mono_assembly_open (path, NULL);
+        assert (assembly);
+        return assembly;
+    }
+    return NULL;
+}
+
+static MonoAssembly*
+mono_assembly_preload_hook (MonoAssemblyName *aname, char **assemblies_path, void* user_data)
+{
+    const char *name = mono_assembly_name_get_name (aname);
+    const char *culture = mono_assembly_name_get_culture (aname);
+    return mono_load_assembly (name, culture);
+}
+
+static unsigned char *
+load_aot_data (MonoAssembly *assembly, int size, void *user_data, void **out_handle)
+{
+    *out_handle = NULL;
+
+    char path [1024];
+    int res;
+
+    MonoAssemblyName *assembly_name = mono_assembly_get_name (assembly);
+    const char *aname = mono_assembly_name_get_name (assembly_name);
+
+    res = snprintf (path, sizeof (path) - 1, "%s/%s.aotdata", bundle_path, aname);
+    assert (res > 0);
+
+    int fd = open (path, O_RDONLY);
+    if (fd < 0) {
+        return NULL;
+    }
+
+    void *ptr = mmap (NULL, size, PROT_READ, MAP_FILE | MAP_PRIVATE, fd, 0);
+    if (ptr == MAP_FAILED) {
+        close (fd);
+        return NULL;
+    }
+
+    close (fd);
+    *out_handle = ptr;
+    return (unsigned char *) ptr;
+}
+
+static void
+free_aot_data (MonoAssembly *assembly, int size, void *user_data, void *handle)
+{
+    munmap (handle, size);
+}
+
+void
+runtime_init_callback ()
+{
+    initialize_runtimeconfig ();
+
+    initialize_appctx_env_variables ();
+
+    register_aot_modules ();
+
+    // register all bundled modules
+
+    mono_set_assemblies_path ((bundle_path && bundle_path[0] != '\0') ? bundle_path : "./");
+
+    mono_jit_set_aot_only (true);
+
+    mono_install_assembly_preload_hook (mono_assembly_preload_hook, NULL);
+
+    // TODO test debug scenario
+#if DEBUG_ENABLED
+    bool wait_for_debugger = false;
+    mono_debug_init (MONO_DEBUG_FORMAT_MONO);
+    if (wait_for_debugger) {
+        char* options[] = { "--debugger-agent=transport=dt_socket,server=y,address=0.0.0.0:55555" };
+        mono_jit_parse_options (1, options);
+    }
+#endif
+
+    mono_install_load_aot_data_hook (load_aot_data, free_aot_data, NULL);
+
+    mono_set_signal_chaining (true);
+
+    mono_jit_init ("mono.self.contained.library"); // Pass in via LibraryBuilder?
+
+%ASSEMBLIES_LOADER%
+}
+
+void
+init_mono_runtime ()
+{
+    bundle_path = monoeg_g_getenv("%ASSETS_PATH%");
```

Comment:
Maybe this should be done later, the library can be loaded very early and embedder doesn't have had the chance to set anything yet in the env variable.

Reply:
If this is the only call into eglib, then maybe you should just use the underlying getenv instead, eglib symbols go through a remap process, that is why they are named monoeg_ so there is a file in runtime eglib.remap.h that pretty much do a #define g_getenv monoeg_g_getenv, and that is optional, so in theory the symbol might not be remapped, but not sure if that is ever used anymore, but the option is still available when building runtime.

Reply:
Good point, moved the environment variable getter into the callback itself as the first thing it does, and leveraged `getenv` directly instead of needing to import `eglib` for just one API.

Yeah.... I realized that was what the symbol was exported as when I couldn't get `g_getenv` working even when adding in a prototype to the beginning of autoinit.c

Should I be adding prototypes for all mono api's? 
`monovm_runtimeconfig_initialize`
`monovm_initialize`
`mono_assembly_open`
`mono_assembly_name_get_name`
`mono_assembly_name_get_culture`
`mono_assembly_get_name`
`mono_assembly_name_get_name`
`mono_set_assemblies_path`
`mono_jit_set_aot_only`
`mono_install_assembly_preload_hook`
`mono_install_load_aot_data_hook`
`mono_set_signal_chaining`
`mono_jit_init`
`mono_set_runtime_init_callback`

Reply:
Should I make a separate `autoinit.h` header as well to hold all of these prototypes?